[1단계 - 상품 관리 기능] 오리(오현서) 미션 제출합니다.

carsago opened the pull request:
안녕하세요 빙봉! 이번 미션의 리뷰 잘 부탁드립니다.

1단계 요구사항 자체가 복잡하거나, 어려운 것은 아니였지만 계속 코드를 짜면서 알아야할 것, 고민해야할 것은 점점 늘어만 가네요 ㅠㅠ

특히, 통합 테스트를 디테일하게 짤려고 노력했는데, 테스트가 자세하고, 세부적일수록 좋다고 생각하지만

```많은 요구사항이나 다양한 제약이 존재하는 실무에서는 통합 테스트를 자세하게 짜는게 현실적으로 어렵지 않을까?``` 라는 생각이 들었습니다.

특히 단위 테스트를 이미 작성했다면, 중복된 테스트를 짠다는(엄밀히는 다를 수 있지만) 느낌이 많이 들었거든요.

그리고 Layer의 분리에 대한 고민이나 API 명세서, 문서화 등에서 추가적인 학습이 필요하겠다는 생각이 들었네요!!

aiaiaiai1 committed:
feat: 상품 저장 기능 추가

aiaiaiai1 committed:
feat: 상품 리스트 출력 기능 추가

aiaiaiai1 committed:
feat: 상품 삭제 기능 추가

aiaiaiai1 committed:
feat: 상품 수정 기능 추가

aiaiaiai1 committed:
feat: 메인 페이지에서 상품 리스트를 출력하는 기능 추가

aiaiaiai1 committed:
feat: Product 유효성 검증 로직 추가

aiaiaiai1 committed:
refactor: Dao에서 Dto가 아닌 domain 객체를 받도록 변경

aiaiaiai1 committed:
feat: ProductRequest의 유효성 검증 추가

aiaiaiai1 committed:
test: ProductDao 테스트 추가

- ProductDao클래스에 save 메서드에서 반환타입 void 에서 Long으로 변경

aiaiaiai1 committed:
feat: ControllerAdvice를 통한 예외 처리 로직 추가

aiaiaiai1 committed:
refactor: 상품 생성시 body에 id를 전달하도록 변경

- product create url을 admin/create -> admin으로 변경

-

aiaiaiai1 committed:
test: 통합 테스트 추가

aiaiaiai1 committed:
docs: 기능 요구 사항 최신화

- 기타 개행 수정

carsago committed:
refactor: ProductDao의 update, delete 결과가 없을 시 예외를 반환하도록 변경

carsago committed:
refactor: 컨트롤러 분리

carsago committed:
test: ProductApiController Test 추가

carsago committed:
test: MainController Test 추가

carsago committed:
test: ProductApiController 통합 테스트 작성

carsago committed:
refactor: Service 메소드에 @Transactional 추가

carsago committed:
docs: README.md 최신화

carsago commented on src/main/java/cart/dto/request/ProductRequest.java:
```diff
@@ -0,0 +1,39 @@
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+
+public class ProductRequest {
```
업데이트와, 생성시의 DTO를 나눌 수도 있지만, 지금 당장은 두 DTO가 동일하다고 여겨져서 굳이 필요하지 않는데 복잡하게 나누고 싶지 않았어요!

aegis1920 replied:
넵 잘하셨습니다! 👍 굳이 필요하지 않으면 나누지 않아도 됩니다!

aegis1920 commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,7 @@
+spring.jackson.property-naming-strategy=SNAKE_CASE
```
해당 속성을 추가해주신 이유가 있을까요?

carsago replied:
이 부분은 처음에 필요한줄 알아서 추가했다가,  지우는 것을 잊어버린 부분이네요 😢

aegis1920 commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE PRODUCT
+(
+    id        INT AUTO_INCREMENT NOT NULL,
```
- auto increment의 경우, 범위가 많이 늘어날 수 있기 때문에 bigint를 사용하기도 합니다.
- unsigned를 추가하면 범위가 더 넓어집니다!

carsago replied:
이 정도는 변경하는 것이 좋겠네요!

aegis1920 commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE PRODUCT
+(
+    id        INT AUTO_INCREMENT NOT NULL,
+    name      VARCHAR NOT NULL,
```
VARCHAR에 사이즈도 넣어주면 좋겠네요

aegis1920 commented on src/main/java/cart/controller/MainController.java:
```diff
@@ -0,0 +1,33 @@
+    public String index(Model model) {
+        List<ProductResponse> products = productService.findAll();
+        model.addAttribute("products", products);
+        return "index";
```
- Model이 반환되는 과정을 말씀해주실 수 있을까요?
- Model이 아닌 ModelAndView를 써보는 건 어떨까요?

carsago replied:
전 HandlerAdapter가 ModelAndView를 바탕으로 View를 매핑하여 반환해주는걸로 알고 있어요.

Controller의 return 값이 어떤 값이던 결국 최종적으로 반환되는 건 ModelAndView 이고요.

예를 들어 Controller에서 다음과 같은 값을 return. 할때..

ModelAndView -> ViewResolver를 바탕으로 양식에 따라 HTML 페이지를 찾아서 매핑하여 반환
String -> 파라미터로 전달받은 Model(비어있어도 상관 없음)과 String 값을 바탕으로 ModelAndView를 만들어서 반환

따라서 Model을 반환하나 ModelAndView를 반환하나 최종 결과는 상관 없는 걸로 알고 있어요.(중간에 매핑하는 과정은 조금 다르지만), 다만 저는 Model을 사용하는 것이 코드가 더 깔끔하다고 느껴져서 이를 활용했었는데요, 구체적으로 이를 명시해주는 ModelAndView를 사용하는 것도 장점이 있겠네요!

aegis1920 replied:
- `Model이 반환되는 과정을 말씀해주실 수 있을까요?` 이 질문은 뷰를 렌더링하는 과정에 대해 한 번 디버깅해보셨으면 좋겠다는 의미에서 질문드렸는데요. 이미 해보신 건지 정말 잘 알고계시네요! 👍 👍 
  - 디버깅하면서 DispatcherServlet.doDispatch() -> ... -> RequestMappingHandlerAdapter.getModelAndView() 이쪽을 간단하게 보시면 좋을 것 같아요!
- `Model이 아닌 ModelAndView를 써보는 건 어떨까요?` 이 질문은 반환하는 방법에는 다양한 방법이 있으니 한 번 작성해보면 어떨까? 하는 취지였습니다ㅎㅎ Controller에서 Model, ModelAndView 뿐만 아니라 다양한 형태로 반환시켜줄 수 있는데요. 시간되시면 org.springframework.web.servlet.View 인터페이스를 상속한 여러 View를 살펴보셔도 도움이 될 것 같습니다 🙂 

> 따라서 Model을 반환하나 ModelAndView를 반환하나 최종 결과는 상관 없는 걸로 알고 있어요.(중간에 매핑하는 과정은 조금 다르지만), 다만 저는 Model을 사용하는 것이 코드가 더 깔끔하다고 느껴져서 이를 활용했었는데요, 구체적으로 이를 명시해주는 ModelAndView를 사용하는 것도 장점이 있겠네요!

맞습니다! 저는 Model보단 ModelAndView가 좀 더 명시적인 코드라고 생각이 들어서 개인적으로 ModelAndView를 더 선호하긴 합니다ㅎㅎ

aegis1920 commented on src/main/java/cart/controller/ProductApiController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<String> create(@RequestBody @Valid ProductRequest productRequest) {
```
- 응답 Body가 없는데 String 타입으로 써주신 이유가 있을까요?
- 성공했으면서 응답 Body가 없다면 어떤 응답 코드를 내려주는 게 좋을까요?

carsago replied:
이 부분은 Void를 보내줬어야 했네요 ㅠㅠ 204 noContent라는 응답 코드도 존재하였구요!
휴일동안 한번 자주 쓰이는 응답 코드들에 대해서 알아보는 것이 좋겠네요!

aegis1920 replied:
REST API 관련 몇 가지 링크 추천드려요! (REST API는 github이 원칙을 잘 지키고 있는 것으로 소문이 자자하니 github docs를 보시는 걸 추천드려요)

- https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28
- https://youtu.be/RP_f5dMoHFc

aegis1920 commented on src/main/java/cart/exception/ProductExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+
+import static java.util.stream.Collectors.joining;
+
+@RestControllerAdvice
```
👍

aegis1920 commented on src/main/java/cart/exception/ProductExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+import static java.util.stream.Collectors.joining;
+
+@RestControllerAdvice
+public class ProductExceptionHandler {
```
- 만약 NullPointerException이 발생하면 어떻게 되나요?
- 발생하는 Exception에 대해서 ExceptionHandler 처리가 필요할까요?
- 모든 Exception에 대해서 처리하려면 어떻게 해줘야 할까요?

carsago replied:
NPE, SQL exception 등의 예외가 발생 안하는 것이 가장 좋겠지만, 발생할 가능성이 없진 않고 이는 클라이언트가 대응이 불가능하니 , 서버 에러(500)를 보내주는 것이 맞다고 느꼈어요. 클라이언트의 실수등으로 발생한 에러(대응가능한 에러)는 4xx 예외를 반환하도록 하고요.

또 이는  Exception.class로 잡아주는 것이 맞다고 느꼈어요.  

그렇게 생각한 이유는 이러한 에러는 보통 작성한 코드에서 문제가 있는 경우도 있을 가능성이 크고, 또 일부러 문제가 있는 코드를 작성하지 않았을 것이라는 점, 서비스가 커져서 외부 API에서도 에러가 발생할 수 있다는 점 등, ```특정한 예외타입으로 예측할 수 없다고 느꼈어요.``` (서비스가 확장되고, 배포했을때  서버 과부하등으로 인한 몇몇 서버 예외등은 따로 처리해줄 수 있을 것 같네요!)

물론 제가 예상치 못한 부분에서 클라이언트의 입력값 실수등을 따로 잡아주지 못해서, 500을 내려주는 경우가 존재할 가능성이 있습니다.

다만 그러한 부분은 어쩔 수 없는 부분이 있다고 느꼈어요. 제가 가능한건 최대한 그러한 상황이 발생하지 않도록, 여러 케이스들을 이해하고, 세심하게 코드를 작성하는 것이 최선이라고 여겼어요. 

실 서비스를 한다면, 로그를 통해 파악하는 방법도 존재하겠네요. 물론 파악하긴 어렵겠지만요 ㅠㅠ


+++
상황을 발생시키지 않도록 노력한다고 했는데 바로 request의 Content-Type 에러를 500으로 보내버렸네요 ㅋㅋㅋ ㅠㅠ, 이미 리뷰를 요청한 상황이라 다음 리뷰에 푸쉬하겠습니다!!

aegis1920 replied:
- 오리의 생각을 잘 말씀해주셨네요! 저도 그렇게 생각합니다! 👍 👍 
- 좀 더 나아가면, 여러 예외에 대해 핸들링 되어있는 [ResponseEntityExceptionHandler](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/servlet/mvc/method/annotation/ResponseEntityExceptionHandler.html)를 상속하는 방법도 있습니다! 한 번 활용해보시는 것도 좋을 것 같아요!

aegis1920 commented on src/main/java/cart/service/ProductService.java:
```diff
@@ -0,0 +1,47 @@
+        return savedId;
+    }
+
+    @Transactional(readOnly = true)
```
readOnly = true 옵션을 준 이유가 있을까요?

carsago replied:
단순 조회용 메소드이기에 옵션을 주었습니다.

물론 ```@Transactional``` 자체를 작성하지 않을 수도 있지만 그 구체적인 이유는 

첫째는, 개인적으로 다른 메소드에서 ```@Transactional``` 을  사용했다면 조회 메소드라도 명시적으로 달아주는 것을 선호하기 때문입니다.(통일성, 가독성 등의 목적)

둘째는, DB 설정에 따라 (현재는 그렇지 않고, 사실 성능을 신경을 상황은 아니지만) master - slave 등이 나누어져있다면 성능 향상을 볼 수 있다는 내용을 봐서 이런 방식을 사용했습니다.

aegis1920 commented on src/main/java/cart/service/ProductService.java:
```diff
@@ -0,0 +1,47 @@
+    @Transactional
+    public void update(Long productId, ProductRequest productRequest) {
+        Product product = productRequest.toEntity();
+        productDao.updateById(productId, product);
```
- deleteById 처럼 메서드 명과 인자 이름(id, productId)을 일관성있게 작성해주시면 좋을 것 같아요!

carsago replied:
맞아요 통일성을 주었어야했는데, 이런 부분이 부족했네요

aegis1920 commented on src/test/java/cart/dao/ProductDaoTest.java:
```diff
@@ -0,0 +1,117 @@
+    private ProductDao productDao;
+
+    private final RowMapper<Product> rowMapper = (resultSet, rowNum) -> {
+        Product product = new Product(
```
inline으로 만들면 더 깔끔하지 않을까요?

carsago replied:
이 부분도 inline 하지 않을 이유가 없었네요!

aegis1920 commented on src/main/java/cart/dao/ProductDao.java:
```diff
@@ -0,0 +1,69 @@
+        hasNoMatchingResult(count);
+    }
+
+    private void hasNoMatchingResult(int count) {
+        if (count == 0) {
+            throw new IllegalArgumentException("존재하지 않는 상품입니다.");
+        }
+    }
```
이 로직은 DAO의 역할일까요?

carsago replied:
이 부분은  ```DAO엔 단순한 쿼리 기능만 있어야하기에 예외 처리를 밑단에서 해줘야 한다.``` 라는 생각도 들고요.

한편으로는 findById를 위한 메소드를 만들고나니, ```예외 상황이 발생한 곳에서 처리하는 것이 더 좋지 않을까?``` 라는 생각도 듭니다.
(findById 메소드는 Try - catch 및 Optional로 활용해서 Service단에서 예외를 처리할 수도 있겠지만, Dao에서 예외를 바꿔서 던져주는 것이 코드가 더 깔끔하고, 유지보수등에서 유리하다고 느꼈어요.)

```어플리케이션 상에서 의도치않게 발생한 예외```와 ```비즈니스적으로 특정한 상황을 예외로 정의시켜서 발생시킨 예외```를 분리해서 보시는 관점일까요? 빙봉의 의견이 궁금합니다!

aegis1920 replied:
> 어플리케이션 상에서 의도치않게 발생한 예외와 비즈니스적으로 특정한 상황을 예외로 정의시켜서 발생시킨 예외를 분리해서 보시는 관점일까요?

분리해서 보되 둘 다 필요하다는 관점인데요. 좀 더 정확히는 DAO 패턴을 사용한 객체(ProductDao)에서 발생한 예외와 비즈니스적으로 발생한 예외(ProductService)를 분리해서 봐야될 것 같습니다. DAO는 Persistent Layer(DB) 접근을 추상화시켜 만든 객체라 지금 작성하신 것처럼 `EmptyResultDataAccessException` 를 캐치해서 다시 던진 `ResourceNotFoundException` 이 적절할 것 같고 이를 다시 ProductService에서 캐치해서 `존재하지 않는 상품입니다` 라는 예외 메시지로 던지는 것이 각 역할(DAO, Service)에 맞는 거라고 생각해요.

`존재하지 않는 상품입니다`라는 예외 메시지는 DAO보단 비즈니스 로직이 적용되는 Service 레이어에서 예외처리가 이뤄저야 할 것 같아서 드린 질문이었습니다 🙂

aegis1920 commented on src/test/java/cart/controller/ProductApiControllerTest.java:
> 특히, 통합 테스트를 디테일하게 짤려고 노력했는데, 테스트가 자세하고, 세부적일수록 좋다고 생각하지만 많은 요구사항이나 다양한 제약이 존재하는 실무에서는 통합 테스트를 자세하게 짜는게 현실적으로 어렵지 않을까? 라는 생각이 들었습니다. 특히 단위 테스트를 이미 작성했다면, 중복된 테스트를 짠다는(엄밀히는 다를 수 있지만) 느낌이 많이 들었거든요.

- 네 맞아요! 다양한 요구사항이 존재하고 복잡한 도메인일수록 통합테스트를 작성하기에 어려워집니다! 👍 
- 지금 작성하신 테스트는 직접 스프링 부트를 띄워서 테스트하는 통합테스트로 생각되는데요. 이렇게 테스트하신 이유가 있으실까요? (서비스 테스트코드는 존재하지 않아서요!) 
  - 만약 중복된 테스트라고 생각되어서 서비스 테스트코드를 작성하지 않았다면 서비스 테스트코드보다 통합테스트를 선호하신 이유도 궁금합니다!

carsago replied:
- 서비스 테스트는 짤게 없는거 같았는데 자고 일어나서 정신이 맑아지니 어떤식으로 짜야할지 떠올라서 추가했어요 ㅎㅎ;

- 중복된 테스트에 대한 제 의견은 ```가능하다면 작성하는 것이 좋다.``` 입니다. 코드가 변경 및 확장될 수 있고, 이러한 부분에서 중복된 테스트라도 존재하는 것이 일종의 뼈대로써, 추가적인 테스트의 작성 의지를 만들어주고, 리소스를 미리 경감시켜준다고 생각해요.

  다만 지금 작성한 일부 테스트는 너무 중복이라 의미가 없는것 처럼 여겨져요.  원인을 생각해보니 지금 당장은 Service layer가 딱히  로직상 존재 의미가 거의 없어서 그렇다는 생각이 드네요.(DAO에 CRUD 쿼리 요청만을 하니까요.)

  지금 당장은 로직이 크지 않아서, 또 이후에 요구사항에 관련 로직이 추가될 수 있다는 점, 학습의 목적 등으로, 이러한 부분을 작성했지만, 이후에 우테코 팀 프로젝트나 실무를 할때 현실적으로 이런 테스트를 다 작성하기엔 리소스가 너무 많이 필요 하다는 생각이 듭니다.
  
   아마 이후에는 통합 테스트와 다른 단위 테스트를 믿고 과감하게 CRUD만 있는 Service Layer의 중복 테스트를 생략하거나, 특정한 부분에는 Service Layer를 제거하는 방법을 고려해볼 것 같아요.

- 통합테스트, 구체적으로 Controller의 통합 테스트를 작성한 이유는 ControllerAdvice 때문이였어요. 처음엔 Controller의 단위 테스트로 작성했는데, ```정상 로직 이외에 예외가 발생했을때, ControllerAdvice가 이를 잡아서 적합한 status code와 메시지를 전달해주는가?```를 테스트하기 위해, 통합 테스트로 작성하였어요.

aegis1920 replied:
> 지금 당장은 로직이 크지 않아서, 또 이후에 요구사항에 관련 로직이 추가될 수 있다는 점, 학습의 목적 등으로, 이러한 부분을 작성했지만, 이후에 우테코 팀 프로젝트나 실무를 할때 현실적으로 이런 테스트를 다 작성하기엔 리소스가 너무 많이 필요 하다는 생각이 듭니다. 아마 이후에는 통합 테스트와 다른 단위 테스트를 믿고 과감하게 CRUD만 있는 Service Layer의 중복 테스트를 생략하거나, 특정한 부분에는 Service Layer를 제거하는 방법을 고려해볼 것 같아요.

👍 👍 오리의 말에 공감하는 부분이 많습니다. 그래서 비즈니스 로직이 들어간 핵심적인 부분을 테스트하는 것이 중요하고, 그 핵심적인 부분을 파악하는 게 중요해요. 테스트도 결국 프로덕션 코드를 테스트하는 거라, 프로덕션에 의존하기 때문에 많은 부분을 테스트할수록 유지보수하는 비용도 그만큼 증가하게 됩니다. 그 비용을 줄이기 위해서 테스트하기 쉬운 프로덕션 코드를 작성해야 것도 있고요. 말씀하신대로 현재는 학습의 목적이라고 생각해주시고 가급적 작성하시는 것을 추천드려요!

> 통합테스트, 구체적으로 Controller의 통합 테스트를 작성한 이유는 ControllerAdvice 때문이였어요. 처음엔 Controller의 단위 테스트로 작성했는데, 정상 로직 이외에 예외가 발생했을때, ControllerAdvice가 이를 잡아서 적합한 status code와 메시지를 전달해주는가?를 테스트하기 위해, 통합 테스트로 작성하였어요.

👍 👍

carsago commented on src/main/java/cart/controller/ProductApiController.java:
```diff
@@ -0,0 +1,45 @@
+
+@RestController
+@RequestMapping("/product")
+public class ProductApiController {
```
이 부분에서 고민한건 ResponseEntity에서 body를 찍어주는 것(create 하거나 update한 Entity의 결과를)인데요.

현재 코드에선 작성하지 않았지고,  상태 코드를 제공하긴 하지만 더 디테일한 정보를 주는것이 이용자들한테 더 좋지 않을까? 라는 점과 테스트하기 더 수월한 코드가 된 다는 생각이 있어서 추가를 고민 중입니다.

혹시 제가 깨닫지 못한 점이나, 다른 컨벤션 같은 부분이 있을까요?

aegis1920 replied:
> 현재 코드에선 작성하지 않았지고, 상태 코드를 제공하긴 하지만 더 디테일한 정보를 주는것이 이용자들한테 더 좋지 않을까? 라는 점과 테스트하기 더 수월한 코드가 된 다는 생각이 있어서 추가를 고민 중입니다.

- 넵! 그 부분은 클라이언트의 요청에 따라 달라집니다! 지금은 클라이언트도 오리가 직접 작성하고 있으니, 편하신대로 작성해보세요!
- 현재 POST요청으로 리소스를 생성할 때 201 응답코드를 내려보내고 있는데요. 이 [링크](https://developer.mozilla.org/ko/docs/Web/HTTP/Status/201) 참고해서 반영해보면 좋을 것 같아요!

aegis1920 commented on src/test/java/cart/controller/ProductApiControllerTest.java:
```diff
@@ -0,0 +1,323 @@
+import org.springframework.jdbc.core.JdbcTemplate;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class ProductApiControllerTest {
```
테스트코드 너무 꼼꼼하게 작성해주셔서 깜작 놀랐습니다 👍

aegis1920 reviewed (changes requested):
안녕하세요 오리! 빙봉입니다 🙂 

질문에 대한 답변 및 코멘트 남겼으니 확인해주세요~

궁금한 점은 언제든 편하게 DM 주세요!

carsago committed:
refactor: Product 테이블 타입 일부 변경

- VARCHAR length 제한
- id int -> bigInt & unsigned
- Inmemory db를 사용하여 JdbcTest가 깨지는 문제 수정

carsago committed:
test: ProductService unit Test 추가

carsago committed:
refactor: update, delete 예외처리를 Service단에서 하도록 변경

carsago committed:
test: Service Test 추가

carsago committed:
refactor: 500 예외 처리 추가

carsago committed:
refactor: MethodArgumentTypeMismatchException을 ExceptionHandelr에서 다루도록 변경

carsago committed:
feat: ProductDao에 findById 추가

carsago committed:
feat: ProductApi에 아이디로 조회하는 기능 추가

- URL 변경 /product -> /products

carsago committed:
reafactor: Product 생성시 Location을 반환하도록 변경

- 개행 변경으로 테스트 코드 가독성 개선

carsago committed:
refactor: CustomException을 활용하도록 변경

carsago committed:
refactor: ModelAndView 반환하도록 변경

carsago commented:
안녕하세요 빙봉 이번 리뷰도 잘부탁드립니다 ~

피드백 주신 부분 위주로 코드를 변경했고, 추가적인 부분은 직접 발생시킨 예외들을 CustomException으로 변경하였습니다.

이전의 미션들에서는  ```CustomException을 만들면 복잡해지고 관리 포인트만 많아질텐데 왜 작성하는거지?, 메시지를 적당히 적어주면 되는거 아닌가?``` 라는 생각을 하였는데요.

스프링을 적용한 미션을 수행하면서 ControllerAdvice를 통해 예외를 핸들링 해주다보니 필요하다고 느껴졌어요! 

이전에는  IllegalArgumentException, NoSuchElementException를 응답코드 4xx로 보내줬었는데, 생각해보니 클라이언트의 잘못된 요청이 아니라, 제가 코드를 잘못 작성시켰을 경우, 라이브러리를 사용하다가 발생하는 경우 등, 클라이언트의 잘못이 아닌 상황에서도 이러한 예외를 4xx로 전달할 수 있겠더군요.! 따라서 이를 방지하기 위해 CustomException 을 활용하였습니다.

또 한가지 질문이 있는데요.

전 현재 예외 발생시 응답값에 오직 제가 정의한 메시지(String message)만 담아서 보내주고 있어요.

그런데 다른 블로그나, 여러 크루들은 ErrorResponse 객체에 상태 코드, 발생 시간등을 담아줘서 보내주는 경우가 있었습니다.

전 클라이언트에게 가능한 디테일한 정보를 주는 것이 좋다고 생각하지만, ```상태 코드나 시간 등은 기본적으로 HTTP response에 담겨져 있는데, 중복된 내용이 필요한가?``` 라는 생각이 들더군요. 제가 파악못한 부분에서 이러한 중복된 값들을 body에 찍어주는 의미가 있을까요?

carsago commented on src/test/java/cart/dao/ProductDaoTest.java:
```diff
@@ -0,0 +1,104 @@
+import org.springframework.boot.test.autoconfigure.jdbc.JdbcTest;
+
+@JdbcTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
```
```@JdbcTest```는 default로 인메모리 DB를 사용하는데, Table id에 unsigned를 사용하는 타 DB 문법을 이용하면 SQL 문법 에러가 발생하더라고요!, 따라서 이 설정을 추가해줬습니다.

aegis1920 replied:
오 맞아요! 👍 이전 크루 리뷰했었을 때 같은 이슈가 있었는데요. 자세한 해결 방법이 궁금하시다면 이쪽 [링크](https://github.com/woowacourse/jwp-racingcar/pull/173#discussion_r1175218467)를 보셔도 좋을 것 같습니다ㅎㅎ

carsago committed:
style: 개행 수정

carsago committed:
refactor: Product imageUrl의 공백값 예외 처리 추가

carsago committed:
docs: API 문서 최신화

carsago commented on src/main/java/cart/domain/Product.java:
```diff
@@ -0,0 +1,70 @@
+        this.name = name;
+        this.price = price;
+        this.imageUrl = imageUrl;
+    }
```
이 부분에서 질문이 하나 또 있는데요. 우테코 프리코스와  Level1 에서 강조했던 원시값 포장에 관한 부분이에요.

그동안 원시값을 포장하는 것의 장점은 학습했지만, 이 값들을 모두 포장하면 클래스가 너무 많아지고 한 곳에서 보기가 어려워진다고 생각해요.

물론 Product라는 클래스가 검증 이외에 별도의 행위가 많이 추가되서 코드가 늘어난다면, 검증 로직을 분리하기 위해 원시값 포장을 고려해보겠지만, 지금은 단순 검증만 존재하기에 나누고 싶지 않았습니다.

빙봉은 이러한 부분에 대해 어떻게 생각하시나요?

aegis1920 replied:
맥락에 따라 중요한 것이 있고 중요하지 않은 것들이 있어서요. 이를 판별하는 능력을 기르고 있는 중이구나 생각하시면 좋을 것 같아요. 실무에서는 무수히 많은 요청과 무수히 많은 필드가 존재하는데요. 이를 모두 포장하고 있지 않습니다.

제가 생각하기에는 여기서 중요하다고 생각되는 필드는 price인데요. 가격이 잘못들어가는 순간 주문과 결제, 환불 등 무수히 많은 부분이 꼬일 수 있기 때문에 price 정도는 원시값 포장을 하는 게 좋을 것 같다라고 생각이 들긴합니다. (특정 상품에 대해서 가격에 대해서 정책이 다를 수 있고, 기본적으로 가격은 마이너스가 안되거나 특정 validation, 불변성 유지 등등..)

aegis1920 commented on src/main/java/cart/domain/Product.java:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        if (price % UNIT_OF_PRICE != 0) {
+            throw new ArgumentNotValidException(String.format("상품의 가격 단위는 %d원 단위입니다.", UNIT_OF_PRICE));
```
기능 요구사항에 100원 단위가 없어서 몰랐는데요. 클라이언트에서 테스트했을 때 submit 시 아무 것도 뜨지 않아서 사용자 입장에서 당황할 수 있을 것 같아요. 클라이언트에서 간단하게 alert 창이라도 뜨게 하는 건 어떨까요?

aegis1920 commented on src/main/java/cart/exception/ExceptionResponse.java:
```diff
@@ -0,0 +1,14 @@
+package cart.exception;
+
+public class ExceptionResponse {
```
> 전 클라이언트에게 가능한 디테일한 정보를 주는 것이 좋다고 생각하지만, `상태 코드나 시간 등은 기본적으로 HTTP response에 담겨져 있는데, 중복된 내용이 필요한가?` 라는 생각이 들더군요. 제가 파악못한 부분에서 이러한 중복된 값들을 body에 찍어주는 의미가 있을까요?

- Response에 담기는 헤더 시간은 따로 set하지 않는 이상 응답할 때 시간을 기준으로 담고 있습니다. ([링크](https://github.com/apache/tomcat/blob/9.0.x/java/org/apache/coyote/http11/Http11Processor.java#L963-L965), tomcat-embed-core:9.0.71에는 같은 라인이 아니지만 코드는 같아서 이해하는데는 상관없을 것 같습니다!)
  - 즉, 에러가 발생한 시각과 Http Response에 담긴 시간은 다릅니다.
  - 클라이언트 측에서 응답하는 시간이 아닌 에러가 발생한 시간이 필요하다면 담을 필요가 있을 것 같고, 아니라면 말씀대로 의미가 없을 것 같습니다.
- 단순히 응답 상태코드로만 표현할 수 있다면 좋겠지만 실제로 그렇게 표현하기가 어렵습니다. 
  - 예를 들어 인증이 되지 않아 실패한 경우, 401 응답 코드(UNAUTHORIZED)를 보낼 수 있는데요.  
  - 만료된 세션 토큰으로 인해 인증에 실패한 경우가 있을 수 있고, 제휴사(구글 등..) 인증을 통해 인증을 시도하는데 구글 쪽에서 에러가 나서 인증에 실패한 경우가 있을 수 있습니다. 이때 HTTP에서 주는 401 응답 코드로만 표현이 가능할까요?
  - HTTP에서 주는 응답 코드로만 표현하기는 힘들어서 보통 401을 prefix로 붙이고 뒤에 개별 코드를 추가해서 에러 코드를 정의하곤 합니다.

aegis1920 reviewed (approved):
안녕하세요 오리! 빙봉입니다 🙂 

피드백 반영하시느라 고생 많으셨습니다! 기존 코멘트에 대한 답변 및 추가 코멘트 드렸으니 2단계에서 반영해주시면 될 것 같아요!

1단계는 여기서 머지하겠습니다!
